core/src/main/scala/org/ensime/indexer/graph/GraphService.scala

```diff
@@ -26,7 +27,63 @@ import org.ensime.util.file._
 import org.ensime.vfs._
 import shapeless.cachedImplicit
 
-// FIXME: these domain objects are remnants of the DatabaseService and should be remodelled
+sealed trait FqnSymbol {
+  def fqn: String
+  def line: Option[Int]
+  def source: Option[String]
+  def declAs: DeclaredAs
+  def access: Access
+
+  def sourceFileObject(implicit vfs: EnsimeVFS): Option[FileObject] = source.map(vfs.vfile)
+}
+
+sealed trait Hierarchy {
+  def toSet: Set[ClassDef] = this match {
+    case cdef @ ClassDef(_, _, _, _, _, _) => Set(cdef)
+    case ClassHierarchy(cdef, refs) => Set(cdef) ++ refs.flatMap(_.toSet)
+  }
+}
+
+final case class ClassHierarchy(aClass: ClassDef, classRefs: Seq[Hierarchy]) extends Hierarchy
+
+sealed trait HierarchyType
```

Comment:
I generally try to follow this rule, but I have no idea how to name this trait without `Type` in it, because it is a **type** that represents a **type** of hierarchy. :)